File: src/main/java/com/milaboratory/mixcr/basictypes/MiXCRMetaInfo.java
/*
 * Copyright (c) 2014-2022, MiLaboratories Inc. All Rights Reserved
 *
 * Before downloading or accessing the software, please read carefully the
 * License Agreement available at:
 * https://github.com/milaboratory/mixcr/blob/develop/LICENSE
 *
 * By downloading or accessing the software, you accept and agree to be bound
 * by the terms of the License Agreement. If you do not want to agree to the terms
 * of the Licensing Agreement, you must not download or access the software.
 */
package com.milaboratory.mixcr.basictypes;

import com.milaboratory.mixcr.assembler.CloneAssemblerParameters;
import com.milaboratory.mixcr.basictypes.tag.TagsInfo;
import com.milaboratory.mixcr.vdjaligners.VDJCAlignerParameters;
import com.milaboratory.primitivio.annotations.Serializable;

import java.util.Objects;
import java.util.function.Function;

/**
 * This class represents common meta-information stored in the headers of vdjca/clna/clns files.
 * The information that is relevant for the downstream analysis.
 */
@Serializable(by = IO.MiXCRMetaInfoSerializer.class)
public final class MiXCRMetaInfo {
    /** Set by used on align step, used to deduce defaults on all downstream steps */
    public final String tagPreset;
    /** Aligner parameters */
    public final TagsInfo tagsInfo;
    /** Aligner parameters */
    public final VDJCAlignerParameters alignerParameters;
    /** Clone assembler parameters */
    public final CloneAssemblerParameters assemblerParameters;

    public MiXCRMetaInfo(String tagPreset, TagsInfo tagsInfo, VDJCAlignerParameters alignerParameters, CloneAssemblerParameters assemblerParameters) {
        this.tagPreset = tagPreset;
        this.tagsInfo = tagsInfo == null ? TagsInfo.NO_TAGS : tagsInfo;
        this.alignerParameters = Objects.requireNonNull(alignerParameters);
        this.assemblerParameters = assemblerParameters;
    }

    public MiXCRMetaInfo withTagInfo(TagsInfo tagsInfo) {
        return new MiXCRMetaInfo(tagPreset, tagsInfo, alignerParameters, assemblerParameters);
    }

    public MiXCRMetaInfo updateTagInfo(Function<TagsInfo, TagsInfo> tagsInfoUpdate) {
        return new MiXCRMetaInfo(tagPreset, tagsInfoUpdate.apply(tagsInfo), alignerParameters, assemblerParameters);
    }

    public MiXCRMetaInfo withAssemblerParameters(CloneAssemblerParameters assemblerParameters) {
        return new MiXCRMetaInfo(tagPreset, tagsInfo, alignerParameters, assemblerParameters);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        MiXCRMetaInfo that = (MiXCRMetaInfo) o;
        return Objects.equals(tagPreset, that.tagPreset) && Objects.equals(tagsInfo, that.tagsInfo) && Objects.equals(alignerParameters, that.alignerParameters) && Objects.equals(assemblerParameters, that.assemblerParameters);
    }

    @Override
    public int hashCode() {
        return Objects.hash(tagPreset, tagsInfo, alignerParameters, assemblerParameters);
    }
}

File: src/main/kotlin/com/milaboratory/mixcr/basictypes/MiXCRMetaInfo.kt
/*
 * Copyright (c) 2014-2022, MiLaboratories Inc. All Rights Reserved
 *
 * Before downloading or accessing the software, please read carefully the
 * License Agreement available at:
 * https://github.com/milaboratory/mixcr/blob/develop/LICENSE
 *
 * By downloading or accessing the software, you accept and agree to be bound
 * by the terms of the License Agreement. If you do not want to agree to the terms
 * of the Licensing Agreement, you must not download or access the software.
 */
package com.milaboratory.mixcr.basictypes

import com.milaboratory.mitool.pattern.search.BasicSerializer
import com.milaboratory.mixcr.assembler.CloneAssemblerParameters
import com.milaboratory.mixcr.basictypes.tag.TagsInfo
import com.milaboratory.mixcr.vdjaligners.VDJCAlignerParameters
import com.milaboratory.primitivio.PrimitivI
import com.milaboratory.primitivio.PrimitivO
import com.milaboratory.primitivio.annotations.Serializable
import com.milaboratory.primitivio.readObjectOptional
import com.milaboratory.primitivio.readObjectRequired
import io.repseq.core.GeneFeature
import io.repseq.core.VDJCLibraryId
import io.repseq.dto.VDJCLibraryData

/**
 * This class represents common meta-information stored in the headers of vdjca/clna/clns files.
 * The information that is relevant for the downstream analysis.
 */
@Serializable(by = MiXCRMetaInfo.SerializerImpl::class)
data class MiXCRMetaInfo(
    /** Set by used on align step, used to deduce defaults on all downstream steps  */
    val tagPreset: String? = null,
    /** Aligner parameters */
    val tagsInfo: TagsInfo = TagsInfo.NO_TAGS,
    /** Aligner parameters */
    val alignerParameters: VDJCAlignerParameters,
    /** Clone assembler parameters  */
    val assemblerParameters: CloneAssemblerParameters? = null,
    /** Library produced by search of alleles */
    val foundAlleles: FoundAlleles?,
    /** If all clones cut by the same feature and cover this feature fully */
    val allFullyCoveredBy: GeneFeatures?
) {
    fun withTagInfo(tagsInfo: TagsInfo): MiXCRMetaInfo =
        copy(tagsInfo = tagsInfo)

    fun updateTagInfo(tagsInfoUpdate: (TagsInfo) -> TagsInfo): MiXCRMetaInfo =
        copy(tagsInfo = tagsInfoUpdate(tagsInfo))

    fun withAssemblerParameters(assemblerParameters: CloneAssemblerParameters): MiXCRMetaInfo =
        copy(assemblerParameters = assemblerParameters)

    fun withAllClonesCutBy(allClonesAlignedBy: Array<GeneFeature>) =
        copy(allFullyCoveredBy = GeneFeatures(allClonesAlignedBy))

    @Serializable(by = FoundAlleles.SerializerImpl::class)
    data class FoundAlleles(
        val libraryName: String,
        val libraryData: VDJCLibraryData
    ) {
        val libraryIdWithoutChecksum: VDJCLibraryId get() = VDJCLibraryId(libraryName, libraryData.taxonId)

        class SerializerImpl : BasicSerializer<FoundAlleles>() {
            override fun write(output: PrimitivO, obj: FoundAlleles) {
                output.writeObject(obj.libraryName)
                output.writeObject(obj.libraryData)
            }

            override fun read(input: PrimitivI): FoundAlleles {
                val libraryName = input.readObjectRequired<String>()
                val libraryData = input.readObjectRequired<VDJCLibraryData>()
                return FoundAlleles(
                    libraryName,
                    libraryData
                )
            }

            override fun isReference(): Boolean = true
        }
    }

    class SerializerImpl : BasicSerializer<MiXCRMetaInfo>() {
        override fun write(output: PrimitivO, obj: MiXCRMetaInfo) {
            output.writeObject(obj.tagPreset)
            output.writeObject(obj.tagsInfo)
            output.writeObject(obj.alignerParameters)
            output.writeObject(obj.assemblerParameters)
            output.writeObject(obj.foundAlleles)
            output.writeObject(obj.allFullyCoveredBy)
        }

        override fun read(input: PrimitivI): MiXCRMetaInfo {
            val tagPreset = input.readObjectOptional<String>()
            val tagsInfo = input.readObjectRequired<TagsInfo>()
            val alignerParameters = input.readObjectRequired<VDJCAlignerParameters>()
            val assemblerParameters = input.readObjectOptional<CloneAssemblerParameters>()
            val foundAlleles = input.readObjectOptional<FoundAlleles>()
            val allFullyCoveredBy = input.readObjectOptional<GeneFeatures>()
            return MiXCRMetaInfo(
                tagPreset,
                tagsInfo,
                alignerParameters,
                assemblerParameters,
                foundAlleles,
                allFullyCoveredBy
            )
        }
    }
}
